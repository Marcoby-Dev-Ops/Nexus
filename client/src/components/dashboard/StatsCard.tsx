import React from 'react';
import PropTypes from 'prop-types';
import Skeleton from '../lib/Skeleton';

/**
 * @interface StatsCardProps
 * @description Props for the StatsCard component.
 * @property {string} title - The KPI title.
 * @property {string | number} value - The main KPI value.
 * @property {number} [delta] - The change from previous period (positive/negative for color).
 * @property {string} [deltaLabel] - Label for the delta (e.g., 'vs last month').
 * @property {React.ReactNode} [icon] - Optional icon to display.
 * @property {number} [progress] - For Task Completion, a value 0-100.
 * @property {React.ReactNode} [sparkline] - Optional sparkline element.
 * @property {boolean} [loading] - Show skeleton loader if true.
 * @property {string} [className] - Additional class names.
 */
export interface StatsCardProps {
  title: string;
  value: string | number;
  delta?: number;
  deltaLabel?: string;
  icon?: React.ReactNode;
  progress?: number;
  sparkline?: React.ReactNode;
  loading?: boolean;
  className?: string;
}

/**
 * @name StatsCard
 * @description KPI tile for dashboard summary.
 * @param {StatsCardProps} props
 * @returns {JSX.Element}
 */
const StatsCard: React.FC<StatsCardProps> = React.memo(
  ({
    title,
    value,
    delta,
    deltaLabel,
    icon,
    progress,
    sparkline,
    loading = false,
    className = '',
  }) => {
    // Delta color logic
    let deltaColor = '';
    if (typeof delta === 'number') {
      deltaColor = delta > 0 ? 'text-success' : delta < 0 ? 'text-destructive' : 'text-muted-foreground';
    }

    return (
      <div className={`bg-card rounded-xl p-6 border border-border flex flex-col gap-2 shadow-sm min-w-0 ${className}`}>
        <div className="flex items-center gap-4 mb-2">
          {icon && <span className="w-8 h-8 flex items-center justify-center bg-muted/40 rounded-full">{icon}</span>}
          <span className="font-semibold text-lg truncate flex-1 min-w-0">{title}</span>
        </div>
        <div className="flex items-end gap-2 min-h-[2.5rem]">
          {loading ? (
            <Skeleton className="h-8 w-32" />
          ) : (
            <span className="text-2xl font-bold tabular-nums truncate">{value}</span>
          )}
          {sparkline && <span className="ml-auto">{sparkline}</span>}
        </div>
        {/* Delta */}
        {typeof delta === 'number' && !loading && (
          <div
            className={`flex items-center text-xs font-semibold ${deltaColor} truncate min-w-0`}
            aria-live="polite"
          >
            {delta > 0 && <><span className="mr-0.5">▲</span><span className="sr-only">increase</span></>}
            {delta < 0 && <><span className="mr-0.5">▼</span><span className="sr-only">decrease</span></>}
            {Math.abs(delta)}%
            {deltaLabel && <span className="ml-1 text-muted-foreground font-normal truncate">{deltaLabel}</span>}
          </div>
        )}
        {/* Progress bar */}
        {typeof progress === 'number' && !loading && (
          <div className="w-full mt-2">
            <div className="w-full h-2 bg-muted/20 rounded-full overflow-hidden">
              <div
                className="h-2 bg-primary rounded-full transition-all duration-500"
                style={{ width: `${progress}%` }}
                role="progressbar"
                aria-valuenow={progress}
                aria-valuemin={0}
                aria-valuemax={100}
              />
            </div>
            <div className="text-xs text-muted-foreground mt-0.5">{progress}% complete</div>
          </div>
        )}
      </div>
    );
  }
);

StatsCard.propTypes = {
  title: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  delta: PropTypes.number,
  deltaLabel: PropTypes.string,
  icon: PropTypes.node,
  progress: PropTypes.number,
  sparkline: PropTypes.node,
  loading: PropTypes.bool,
  className: PropTypes.string,
};

/**
 * To show a mini sparkline in a KPI card, pass a ReactNode to the sparkline prop.
 * Example:
 * <StatsCard ... sparkline={<MySparklineComponent data={...} />} />
 */

export default StatsCard; 
