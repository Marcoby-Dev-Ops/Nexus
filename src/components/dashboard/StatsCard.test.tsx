/**
 * @file StatsCard.test.tsx
 * @description Unit and snapshot tests for the StatsCard dashboard component.
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import StatsCard from './StatsCard';

describe('StatsCard', () => {
  it('renders title and value', () => {
    render(<StatsCard title="Tasks" value={42} />);
    expect(screen.getByText('Tasks')).toBeInTheDocument();
    expect(screen.getByText('42')).toBeInTheDocument();
  });

  it('renders icon if provided', () => {
    render(<StatsCard title="Tasks" value={42} icon={<span data-testid="icon">icon</span>} />);
    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('renders delta and deltaLabel with correct color (positive)', () => {
    render(<StatsCard title="Revenue" value={1000} delta={5} deltaLabel="vs last month" />);
    const delta = screen.getByText(/5%/);
    expect(delta).toHaveClass('text-success');
    expect(screen.getByText('vs last month')).toBeInTheDocument();
  });

  it('renders delta with correct color (negative)', () => {
    render(<StatsCard title="Revenue" value={1000} delta={-3} />);
    const delta = screen.getByText(/3%/);
    expect(delta).toHaveClass('text-destructive');
  });

  it('renders progress bar if progress is provided', () => {
    render(<StatsCard title="Progress" value={80} progress={80} />);
    expect(screen.getByRole('progressbar')).toHaveAttribute('aria-valuenow', '80');
    expect(screen.getByText('80% complete')).toBeInTheDocument();
  });

  it('renders sparkline if provided', () => {
    render(<StatsCard title="Revenue" value={1000} sparkline={<span data-testid="sparkline">spark</span>} />);
    expect(screen.getByTestId('sparkline')).toBeInTheDocument();
  });

  it('shows skeleton when loading', () => {
    const { container } = render(<StatsCard title="Revenue" value={1000} loading />);
    // Look for the skeleton element by className
    expect(container.querySelector('.h-8.w-32')).toBeInTheDocument();
  });

  it('applies custom className', () => {
    const { container } = render(<StatsCard title="Revenue" value={1000} className="test-class" />);
    expect(container.firstChild).toHaveClass('test-class');
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<StatsCard title="Revenue" value={1000} delta={5} progress={80} sparkline={<span />} icon={<span />} />);
    expect(asFragment()).toMatchSnapshot();
  });
}); 