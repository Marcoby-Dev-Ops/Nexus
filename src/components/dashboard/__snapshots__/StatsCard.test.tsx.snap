// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-card rounded-xl p-6 border border-border flex flex-col gap-2 shadow-sm min-w-0 "
  >
    <div
      class="flex items-center gap-4 mb-2"
    >
      <span
        class="w-8 h-8 flex items-center justify-center bg-muted/40 rounded-full"
      >
        <span />
      </span>
      <span
        class="font-semibold text-lg truncate flex-1 min-w-0"
      >
        Revenue
      </span>
    </div>
    <div
      class="flex items-end gap-2 min-h-[2.5rem]"
    >
      <span
        class="text-2xl font-bold tabular-nums truncate"
      >
        1000
      </span>
      <span
        class="ml-auto"
      >
        <span />
      </span>
    </div>
    <div
      aria-live="polite"
      class="flex items-center text-xs font-semibold text-success truncate min-w-0"
    >
      <span
        class="mr-0.5"
      >
        â–²
      </span>
      <span
        class="sr-only"
      >
        increase
      </span>
      5%
    </div>
    <div
      class="w-full mt-2"
    >
      <div
        class="w-full h-2 bg-muted/20 rounded-full overflow-hidden"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="80"
          class="h-2 bg-primary rounded-full transition-all duration-500"
          role="progressbar"
          style="width: 80%;"
        />
      </div>
      <div
        class="text-xs text-muted-foreground mt-0.5"
      >
        80% complete
      </div>
    </div>
  </div>
</DocumentFragment>
`;
